coreos_prometheus/Tiltfile

```diff
@@ -0,0 +1,192 @@
+        replace_target(f, target)
+
+def setup_monitoring():
+    k8s_yaml(setupManifests)
+
+    # Wait until the CRDs are ready.
+    local_resource(
+        'prometheus-crds-ready',
+        cmd=' && '.join([('kubectl wait --for=condition=Established crd %s' % name(c)) for c in crds]),
+        resource_deps=['prometheus-crds'],
+    )
+
+    k8s_yaml(mainManifests)
+
+    crdResources = {}
+    for yDoc in mainManifests:
+        r = read_yaml(yDoc)
+        if r['kind'] in crdTypes:
+            rname = name(r)
+            if rname not in crdResources:
+                crdResources[rname] = []
+
+            crdResources[rname].append('%s:%s' % (rname, r['kind']))
+
+    k8s_resource(
+        new_name='prometheus-crds',
+        objects = [('%s' % name(c)) for c in crds],
+        resource_deps=['prometheus-operator'],
+    )
+
+    resources = {
+        'grafana': {
+            'port_forwards': ["3000"],
+        },
+        'node-exporter': {},
+        'kube-state-metrics': {},
+        'prometheus': {
+            'new_name': 'prometheus',
+            'objects': ['prometheus-k8s:service'],
+            'crdResources': ['prometheus', 'k8s'],
+            'extra_pod_selectors': [{'prometheus': 'k8s'}],
+            'port_forwards': ["9090"],
+        },
+        'alertmanager': {
+            'new_name': 'alertmanager',
+            'objects': ['alertmanager-main:service'],
+            'crdResources': ['main'],
+            'extra_pod_selectors': [{'alertmanager': 'main'}],
+            'port_forwards': ["9093"],
+        }
+    }
+
+    for rname, resource in resources.items():
+        args = []
+
+        if 'new_name' not in resource:
+            args.append(rname)
+
+        for crd in resource.pop('crdResources', []):
+            resource['objects'].extend(crdResources.pop(crd, []))
+
+        k8s_resource(*args, resource_deps=['prometheus-crds'], **resource)
+
+    remainingCrdResources = []
+    for res in crdResources.values():
+        remainingCrdResources.extend(res)
+
+    k8s_resource(
+        new_name='uncategorized-prometheus-resources-requiring-crds',
+        objects=remainingCrdResources,
+        resource_deps=['prometheus-crds'],
+    )
+
+def list_prometheus_crd_types():
+    return crdTypes
+
+def get_prometheus_dependencies():
+    return ['prometheus-crds']
+
+def get_prometheus_resources(yaml, name):
+    results = []
+    for crd in list_prometheus_crd_types():
+        found, _ = filter_yaml(yaml, name="^%s$" % name, kind="^%s$" % crd)
+        if found:
+            results.append('%s:%s' % (name, crd))
+
+    return results
+
+
+# To allow resources that require CRDs created by this extension to be selectively
+# pruned from other components if deploying of prometheus is disabled by the
+# file loading this one, need to always download and unpack the required files
+# and perform a limited amount of parsing.
+
+basedir = download_files()
+patch_prometheus(basedir)
```

Comment:
discovered a bug here locally, on reload of the Tiltfile can end up in a constant loop. I'll need to test this code to be sure, but we've had to modify to ensure the files are only patched once.

I was previously thinking that if someone modified the files allowing to detect that they were modified and re-patching would be a good idea, but it doesn't quite work that way.

Below all the relevant files are read, and captured, so during a reload when the patch_prometheus is called to replace the files again, it ends up modifying files that are being watched resulting in a reload, which patches the files again. Either the patch needs to check if the files are different before replacing, or needs to be run once.

Reply:
ya! a few extensions have this problem. the next version of Tilt is going to have a `watch_settings(ignore=...)` setting, so that any extension can tell tilt to ignore files. In the meantime, you can probably add a little note to the README like:

add `tilt_modules/coreos_prometheus/...` to your `.tiltignore`